### DISCUSSION TITLE
X-ray Upper Limits Divide by Zero Error & Negative Yerr

### DISCUSSION THREAD

**Discussion by Vasilisa-Malenkiy** [mboquien/cigale#101] [Q&A]
**DIVIDE BY ZERO ERROR:** I'm trying to add an X-ray upper limit to my input file. I do this by setting the error to a negative number and enabling lim_flag = full. However, two RunTimeWarnings pop up when the program estimates the physical properties while executing 'pcigale run.' Both point to the file defined by this path - cigale-v2022.1/pcigale/analysis_modules/pdf_analysis/utils.py and they are divide by zero errors on lines 93 and 310. I did some debugging and narrowed it down to two potential options - 1) obslim_err or obs_error are being rounded down to zero somehow (when I set the upper limit error to a larger number like -0.1 versus -0.000000000839 everything works fine. Similarly, adding a small constant onto obslim_err resolves the issue but it's an unscientific fix.) or 2) the variable s is registering as NaN and throwing off the calculation. Does anyone have any thoughts on what else could be causing this or how to fix these errors so I can add the xray upper limits?

I've attached two input files - the first is the one I eventually hope to run, the second is one source I'm focusing on for the purposes of debugging - the utils.py file, and a screenshot of the divide by zero error.
[goods-s.2X.txt](https://github.com/user-attachments/files/16313780/goods-s.2X.txt)
[210625.txt](https://github.com/user-attachments/files/16313783/210625.txt)
[utils.pdf](https://github.com/user-attachments/files/16313867/utils.pdf)
<img width="568" alt="Screenshot 2024-07-19 at 11 07 43 AM" src="https://github.com/user-attachments/assets/50e5e0c1-a70a-4a56-9c8c-840b3284228e">

Thank you so much!

**Comment by Guang91** [comment]
have you tried using -8.39e-10 insted of -0.000000000839?

**Reply by Vasilisa-Malenkiy** [reply]
Yes, and unfortunately, the errors still persist.

**Comment by Guang91** [comment]
Could you provide your pcigale.ini file? I will try to see if I can reproduce your error.

**Reply by Vasilisa-Malenkiy** [reply]
[pcigale.ini.pdf](https://github.com/user-attachments/files/16325298/pcigale.ini.pdf)

**Comment by Guang91** [comment]
I've tried out your data. I also have the warning message, but the output looks fine. 

I think the warning is because some models are ruled out because their template fluxes are too much higher than the upper limits and the code can't calculate the exact value of chi2 (so it simply set chi2=nan).

Is there anything wrong in your output? If not, I think you can just go ahead.

**Reply by Vasilisa-Malenkiy** [reply]
Good to know! I ran into a separate negative y_err issue which I resolved by adding abs() around the appropriate variables and now that I'm able to see the output, it all looks fine. Thanks.
